File: backend/ai/utils/textUtils.js
/**
 * Text Processing Utilities for AI System
 * Handles text chunking, cleaning, and extraction
 */

const AI_CONFIG = require('../config/aiConfig');
const AILogger = require('./logger');

const logger = new AILogger('TextUtils');

/**
 * Split text into chunks with overlap
 */
function chunkText(text, chunkSize = null, chunkOverlap = null) {
  chunkSize = chunkSize || AI_CONFIG.chunking.chunkSize;
  chunkOverlap = chunkOverlap || AI_CONFIG.chunking.chunkOverlap;

  if (!text || text.length === 0) {
    return [];
  }

  const chunks = [];
  let start = 0;

  while (start < text.length) {
    let end = start + chunkSize;

    // Try to break at a natural boundary
    if (end < text.length) {
      // Look for separators
      for (const separator of AI_CONFIG.chunking.separators) {
        const lastIndex = text.lastIndexOf(separator, end);
        if (lastIndex > start) {
          end = lastIndex + separator.length;
          break;
        }
      }
    }

    const chunk = text.substring(start, end).trim();
    if (chunk.length > 0) {
      chunks.push(chunk);
    }

    // Move start position with overlap
    start = end - chunkOverlap;
  }

  return chunks;
}

/**
 * Clean text content
 */
function cleanText(text) {
  if (!text) return '';

  // Remove extra whitespace
  text = text.replace(/\s+/g, ' ').trim();

  // Remove control characters
  text = text.replace(/[\x00-\x1F\x7F]/g, '');

  // Normalize line endings
  text = text.replace(/\r\n/g, '\n');

  return text;
}

/**
 * Extract summary from text
 */
function extractSummary(text, maxLength = 500) {
  if (!text) return '';

  const cleaned = cleanText(text);
  const sentences = cleaned.match(/[^.!?]+[.!?]+/g) || [cleaned];

  let summary = '';
  for (const sentence of sentences) {
    if ((summary + sentence).length <= maxLength) {
      summary += sentence;
    } else {
      break;
    }
  }

  return summary.trim() || cleaned.substring(0, maxLength);
}

/**
 * Extract keywords from text
 */
function extractKeywords(text, maxKeywords = 10) {
  if (!text) return [];

  const cleaned = cleanText(text).toLowerCase();

  // Common stop words
  const stopWords = new Set([
    'the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for',
    'of', 'with', 'by', 'from', 'is', 'are', 'was', 'were', 'be', 'been',
    'being', 'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would',
    'could', 'should', 'may', 'might', 'must', 'can', 'this', 'that',
    'these', 'those', 'i', 'you', 'he', 'she', 'it', 'we', 'they',
  ]);

  // Extract words
  const words = cleaned.match(/\b\w+\b/g) || [];

  // Count word frequency
  const wordFreq = {};
  for (const word of words) {
    if (word.length > 3 && !stopWords.has(word)) {
      wordFreq[word] = (wordFreq[word] || 0) + 1;
    }
  }

  // Sort by frequency
  const keywords = Object.entries(wordFreq)
    .sort((a, b) => b[1] - a[1])
    .slice(0, maxKeywords)
    .map(([word]) => word);

  return keywords;
}

/**
 * Extract code blocks from text
 */
function extractCodeBlocks(text) {
  const codeBlocks = [];
  const codeRegex = /```[\s\S]*?```/g;
  let match;

  while ((match = codeRegex.exec(text)) !== null) {
    const block = match[0].replace(/```/g, '').trim();
    if (block.length > 0) {
      codeBlocks.push(block);
    }
  }

  return codeBlocks;
}

/**
 * Extract links from text
 */
function extractLinks(text) {
  const links = [];
  const linkRegex = /https?:\/\/[^\s]+/g;
  let match;

  while ((match = linkRegex.exec(text)) !== null) {
    links.push(match[0]);
  }

  return links;
}

/**
 * Truncate text to max length
 */
function truncateText(text, maxLength = 1000, suffix = '...') {
  if (!text) return '';
  if (text.length <= maxLength) return text;

  return text.substring(0, maxLength - suffix.length) + suffix;
}

/**
 * Calculate text similarity (simple Jaccard similarity)
 */
function calculateTextSimilarity(text1, text2) {
  if (!text1 || !text2) return 0;

  const words1 = new Set(text1.toLowerCase().match(/\b\w+\b/g) || []);
  const words2 = new Set(text2.toLowerCase().match(/\b\w+\b/g) || []);

  const intersection = new Set([...words1].filter(x => words2.has(x)));
  const union = new Set([...words1, ...words2]);

  return intersection.size / union.size;
}

module.exports = {
  chunkText,
  cleanText,
  extractSummary,
  extractKeywords,
  extractCodeBlocks,
  extractLinks,
  truncateText,
  calculateTextSimilarity,
};

File: python-ai/utils/textUtils.py
"""
Text Processing Utilities for Python AI Service
"""

from typing import List
from config.aiConfig import CHUNKING_CONFIG


def chunk_text(text: str, chunk_size: int = None, chunk_overlap: int = None) -> List[str]:
    """Split text into chunks with overlap"""
    chunk_size = chunk_size or CHUNKING_CONFIG['chunk_size']
    chunk_overlap = chunk_overlap or CHUNKING_CONFIG['chunk_overlap']

    if not text or len(text) == 0:
        return []

    chunks = []
    start = 0

    while start < len(text):
        end = start + chunk_size

        # Try to break at a natural boundary
        if end < len(text):
            for separator in CHUNKING_CONFIG['separators']:
                last_index = text.rfind(separator, start, end)
                if last_index > start:
                    end = last_index + len(separator)
                    break

        chunk = text[start:end].strip()
        if chunk:
            chunks.append(chunk)

        start = end - chunk_overlap

    return chunks


def clean_text(text: str) -> str:
    """Clean text content"""
    if not text:
        return ""

    # Remove extra whitespace
    text = ' '.join(text.split())

    # Remove control characters
    text = ''.join(char for char in text if ord(char) >= 32 or char in '\n\t')

    # Normalize line endings
    text = text.replace('\r\n', '\n')

    return text


def extract_summary(text: str, max_length: int = 500) -> str:
    """Extract summary from text"""
    if not text:
        return ""

    cleaned = clean_text(text)
    sentences = [s.strip() for s in cleaned.split('.') if s.strip()]

    summary = ""
    for sentence in sentences:
        if len(summary) + len(sentence) + 1 <= max_length:
            summary += sentence + ". "
        else:
            break

    return summary.strip() or cleaned[:max_length]


def extract_keywords(text: str, max_keywords: int = 10) -> List[str]:
    """Extract keywords from text"""
    if not text:
        return []

    cleaned = clean_text(text).lower()

    # Common stop words
    stop_words = {
        'the', 'a', 'an', 'and', 'or', 'but', 'in', 'on', 'at', 'to', 'for',
        'of', 'with', 'by', 'from', 'is', 'are', 'was', 'were', 'be', 'been',
        'being', 'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would',
        'could', 'should', 'may', 'might', 'must', 'can', 'this', 'that',
        'these', 'those', 'i', 'you', 'he', 'she', 'it', 'we', 'they',
    }

    # Extract words
    words = cleaned.split()

    # Count word frequency
    word_freq = {}
    for word in words:
        if len(word) > 3 and word not in stop_words:
            word_freq[word] = word_freq.get(word, 0) + 1

    # Sort by frequency
    keywords = sorted(word_freq.items(), key=lambda x: x[1], reverse=True)
    return [word for word, _ in keywords[:max_keywords]]


def truncate_text(text: str, max_length: int = 1000, suffix: str = "...") -> str:
    """Truncate text to max length"""
    if not text:
        return ""
    if len(text) <= max_length:
        return text

    return text[:max_length - len(suffix)] + suffix
